### ISSUE TITLE
twisted.runner inetd stuff leaves processes running

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7187]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#425|
|Type|defect|
|Created|2003-12-10 18:26:50Z|

<details><summary>Searchable metadata</summary>

```
trac-id__425 425
type__defect defect
reporter__radix radix
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1071080810000000 1071080810000000
changetime__1469600715760171 1469600715760171
version__ 
owner__ 
cc__radix cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I have twisted.runner delegating TCP connections to an ssh command, and the ssh
instances are staying around even after connections close. I don't know if it's
really runner's responsibility to actually *kill* these processes (what does
inetd do?), or if it just needs to close down FDs properly, or something, but
it's not. Here's my entire script/configuration:

from twisted.runner import inetdconf, inetd
from twisted.application import internet, service

serv = inetdconf.InetdService('name!', 143, 
                              'stream', 'tcp', 'nowait',
                              1000, 1000, 
                              'ssh', 
    ['ssh', 'radix@...', '~/bin/courier-imap'])

fact = inetd.InetdFactory(serv)
application = service.Application('process_proxy',
                                  uid=1000,gid=1000)
internet.TCPServer(143, fact).setServiceParent(application)

Here's a traceback I'm receiving occasionally:

2003/12/10 07:22 EST [twisted.runner.inetd.InetdFactory] Traceback (most recent
call last):
          File "./twisted/python/log.py", line 65, in callWithLogger
            
          File "./twisted/python/log.py", line 52, in callWithContext
            
          File "/home/chris/Projects/Twisted/twisted/python/context.py", line
32, in callWithContext
            return func(*args,**kw)
          File "./twisted/internet/default.py", line 527, in _doReadOrWrite
            
        --- &lt;exception caught here> ---
          File "/home/chris/Projects/Twisted/twisted/internet/tcp.py", line 635,
in doRead
            protocol.makeConnection(transport)
          File "/home/chris/Projects/Twisted/twisted/internet/protocol.py", line
310, in makeConnection
            self.connectionMade()
          File "/home/chris/Projects/Twisted/twisted/runner/inetd.py", line 111,
in connectionMade
            service.user, service.group, self.transport)
          File "/home/chris/Projects/Twisted/twisted/runner/inetd.py", line 89,
in forkPassingFD
            stderr = os.fdopen(2, 'w')
        exceptions.OSError: [Errno 9] Bad file descriptor
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
(Wow -- I didn't realise inetd had any users! ;)

That traceback is a bit wierd -- I think for some reason I don't understand,
os.fdopen will raise Bad File Descriptor when you reopen the same fd enough
times.  I guess in this case, the code could just use sys.stderr anyway -- it
looks to me like the fdopen approach is a hangover from cannibilising the code
from twisted.internet.process.

I suspect a better fix would be to somehow refactor twisted.internet.process to
allow inetd to use that.  There's also certainly been bugfixes to t.i.p that
haven't been applied to inetd (at a glance I see t.i.p doing euid/egid stuff,
which inetd completely ignores).  On the other hand, inetd's needs are overall
simpler and more specialised than t.i.p's, so perhaps polluting t.i.p with
inetd's use-case is a bad idea.

Anyway, here's a trivial (and untested!) diff against inetd.py that should fix
that traceback:

diff -u -r1.2 inetd.py
--- twisted/runner/inetd.py     7 Oct 2003 17:58:36 -0000       1.2
+++ twisted/runner/inetd.py     11 Dec 2003 01:15:16 -0000
@@ -86,7 +86,7 @@
             os.execvpe(exe, args, env)
         except:
             # If anything goes wrong, just die.
-            stderr = os.fdopen(2, 'w')
+            from sys import stderr
             stderr.write('Unable to spawn child:\n')
             traceback.print_exc(file=stderr)

This buglet would cause processes to leak, so hopefully this resolves your
problem.  t.runner.inetd makes no attempt to kill child processes once spawned,
though (which I think is probably correct; inetd has no way in general to know
if the child is really done or not, even if the socket is closed the child might
need to flush its buffers to disk, or something).

Let me know how it goes :)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Andrew Bennetts [Twisted issue tracker] wrote:
> Andrew Bennetts &lt;andrew-twistedbugs@...> added the comment:
> 
> (Wow -- I didn't realise inetd had any users! ;)

Yeah, I'm using it to connect to an imap server on pyramid which I don't 
want to run on TCP. I have a t.runner server on my workstation serving 
port 143 as "ssh radix@... ~/bin/courier-imap" :-)

> Anyway, here's a trivial (and untested!) diff against inetd.py that should fix
> that traceback:

I've applied this patch.

> This buglet would cause processes to leak, so hopefully this resolves your
> problem.

However, the leaking processes have no relation to the occurences of 
this traceback. It seems that the processes are *never* dying: every new 
connection starts a new one, and it never gets destroyed.

>  t.runner.inetd makes no attempt to kill child processes once spawned,
> though (which I think is probably correct; inetd has no way in general to know
> if the child is really done or not, even if the socket is closed the child might
> need to flush its buffers to disk, or something).

So, maybe there's something wrong with the way it's closing file 
descriptors when the socket gets disconnected? I assume ssh would shut 
down when its STDIO FDs have been closed...?

Anyway, I've applied it, and I'll tell you how it goes.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok, spiv helped me fix this bug on IRC. We patched process.py to reap the
processes that inetd was starting. Also, all bad FD bugs were fixed, however,
I'm going to open a new issue about default behavior of stderr with inetd.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|GitHub commented|
|-|-|

In changeset 99137f53a8a6bd1d8bffa5202373236162daf94e
```
#!CommitTicketReference repository="" revision="99137f53a8a6bd1d8bffa5202373236162daf94e"
Merge pull request #425 from twisted/8672-acabhishek942-bytes

Author: acabhishek942
Reviewer: rodrigc
Fixes: #8672

Use bytes instead of string in twisted.internet._sslverify
```
